From f24bad78f475e16163636c8c17f4b667791a04a4 Mon Sep 17 00:00:00 2001
From: Anson Huang <b20788@freescale.com>
Date: Wed, 3 Aug 2011 09:47:32 +0800
Subject: ENGR00154211 [MX6]Add workaround for wdog errata

Errata number:TKT039676

WDOG sw reset is generated by writing to its
control register. WDOG's reset is activated by
ipg_clk_s, and is de-activated (later) by a
synchronized CKIL (32KHz clock). On the other
hand SRC samples the WDOG reset with an
unsynchronized CKIL clock. If the write to WDOG
control register happens between the edges of
unsynchronized and synchronized CKIL clocks SRC
will miss the wdog reset pulse.

Workaround: write wdog control register twice.

Signed-off-by: Anson Huang <b20788@freescale.com>
---
 arch/arm/plat-mxc/system.c |    4 ++++
 1 file changed, 4 insertions(+)

diff --git a/arch/arm/plat-mxc/system.c b/arch/arm/plat-mxc/system.c
index 383013c..cadf3bb 100644
--- a/arch/arm/plat-mxc/system.c
+++ b/arch/arm/plat-mxc/system.c
@@ -43,6 +43,10 @@ void arch_reset(char mode, const char *cmd)
 	/* wait for reset to assert... */
 	wcr_enable = (1 << 2);
 	__raw_writew(wcr_enable, wdog_base);
+	/* errata TKT039676, SRS bit may be missed when
+	SRC sample it, need to write the wdog controller
+	twice to avoid it */
+	__raw_writew(wcr_enable, wdog_base);
 
 	/* wait for reset to assert... */
 	mdelay(500);
-- 
1.7.9.5

